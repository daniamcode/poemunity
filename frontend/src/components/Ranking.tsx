import { useState, useEffect } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import './Ranking.scss'
import { getRanking } from '../utils/getRanking'
import CircularProgress from './CircularIndeterminate'
import {
    RANKING_TITLE,
    RANKING_SUBTITLE,
    RANKING_POETS_TITLE,
    RANKING_POINTS_TITLE,
    POEM_POINTS,
    LIKE_POINTS
} from '../data/constants'
import { useSelector } from 'react-redux'
import { useAppDispatch, RootState } from '../redux/store'
import { getRankingAction } from '../redux/actions/poemsActions'
import { Poem } from '../typescript/interfaces'

const useStyles = makeStyles({
    table: {
        minWidth: 50
    }
})

export default function Ranking() {
    interface RankingStates {
        poems: Poem[]
        rank: object
    }
    const classes = useStyles()

    const [poems, setPoems] = useState<RankingStates['poems']>([])

    const [rank, setRank] = useState<RankingStates['rank']>({})

    // Redux
    const dispatch = useAppDispatch()

    const { rankingQuery } = useSelector((state: RootState) => state)

    useEffect(() => {
        dispatch(
            getRankingAction({
                params: {
                    origin: 'user'
                }
            })
        )
    }, [dispatch])

    useEffect(() => {
        if (rankingQuery?.item) {
            setPoems(rankingQuery?.item)
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [JSON.stringify(rankingQuery?.item)])

    useEffect(() => {
        if (poems) {
            setRank(getRanking(poems, POEM_POINTS, LIKE_POINTS))
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [JSON.stringify([poems, POEM_POINTS, LIKE_POINTS])])

    if (rankingQuery.isFetching) {
        return <CircularProgress data-test='ranking__loading' />
    }

    return (
        <main className='ranking'>
            <h3 className='ranking__title'>{RANKING_TITLE}</h3>
            <h5 className='ranking__subtitle'>{RANKING_SUBTITLE}</h5>
            <TableContainer className='ranking__body' component={Paper}>
                <Table className={classes.table} aria-label='simple table'>
                    <TableHead>
                        <TableRow>
                            <TableCell align='center'>{RANKING_POETS_TITLE}</TableCell>
                            <TableCell align='center'>{RANKING_POINTS_TITLE}</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {Object.keys(rank).map((element, index) => (
                            <TableRow key={index}>
                                <TableCell
                                    className='ranking__picture-container'
                                    align='center'
                                    component='th'
                                    scope='row'
                                >
                                    <div className='ranking__picture-wrap'>
                                        <img className='ranking__picture' src={rank[element]?.picture} />
                                        <p className='ranking__picture-description'>{rank[element]?.author}</p>
                                    </div>
                                </TableCell>
                                <TableCell className='ranking__number' align='center'>
                                    {rank[element]?.points}
                                </TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
        </main>
    )
}
