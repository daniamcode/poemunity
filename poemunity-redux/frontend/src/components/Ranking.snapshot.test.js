import React from 'react'
import renderer from 'react-test-renderer'
import Ranking from './Ranking'
import { BrowserRouter } from 'react-router-dom'
import {
  QueryClient,
  QueryClientProvider,
} from "react-query"

function renderRanking (arg) {
  const defaultProps = {
    match: {
      params: {}
    }
  }

  const props = { ...defaultProps, ...arg }
  const queryClient = new QueryClient();

  return renderer.create(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <Ranking {...props} />
      </BrowserRouter>
    </QueryClientProvider>
  )
}

describe('Ranking', () => {
  let RankingTree

  beforeEach(async () => {
    RankingTree = renderRanking()
  })

  it('should match without id', async () => {
    expect(RankingTree).toMatchSnapshot()
  })
})
