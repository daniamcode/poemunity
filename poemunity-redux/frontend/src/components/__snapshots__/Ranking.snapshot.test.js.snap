// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ranking should match without id 1`] = `
<main
  className="ranking"
>
  <h3
    className="ranking__title"
  >
    Poets ranking
  </h3>
  <h5
    className="ranking__subtitle"
  >
    3 points per poem, 1 per like
  </h5>
  <div
    className="MuiPaper-root MuiTableContainer-root ranking__body MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      aria-label="simple table"
      className="MuiTable-root makeStyles-table-1"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            Poets
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            Points
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</main>
`;
